src/controllerclientimpl.cpp

```diff
@@ -1575,30 +1619,40 @@ void ControllerClientImpl::_UploadFileToController(FILE* fd, const std::string&
     //printf("http code: %d, Speed: %.3f bytes/sec during %.3f seconds\n", http_code, speed_upload, total_time);
 }
 
-void ControllerClientImpl::_UploadFileToControllerViaForm(const std::string& file, const std::string& uri)
-{
-    MUJIN_LOG_DEBUG(str(boost::format("upload %s")%uri))
 
-    // the dest filename of the upload is determined by stripping the leading _basewebdavuri
-    if( uri.size() < _basewebdavuri.size() || uri.substr(0,_basewebdavuri.size()) != _basewebdavuri ) {
-        throw MUJIN_EXCEPTION_FORMAT("trying to upload a file outside of the webdav endpoint is not allowed: %s", uri, MEC_HTTPServer);
-    }
-    std::string filename = uri.substr(_basewebdavuri.size());
 
-    const std::string& endpoint = _baseuri + "fileupload";
+void ControllerClientImpl::_UploadFileToControllerViaForm(std::istream& inputStream, const std::string& filename, const std::string& endpoint)
+{
     CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_URL, NULL, endpoint.c_str());
     _buffer.clear();
     _buffer.str("");
     CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_WRITEFUNCTION, NULL, _WriteStringStreamCallback);
     CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_WRITEDATA, NULL, &_buffer);
 
+    inputStream.seekg(0, std::ios::end);
+    if(inputStream.fail()) {
+        throw MUJIN_EXCEPTION_FORMAT("failed to seek inputStream (%s) to get the length", filename, MEC_InvalidArguments);
+    }
+    size_t contentLength = inputStream.tellg();
+    if(inputStream.fail()) {
+        throw MUJIN_EXCEPTION_FORMAT("failed to tell the length of inputStream (%s)", filename, MEC_InvalidArguments);
+    }
+    inputStream.seekg(0, std::ios::beg);
+    if(inputStream.fail()) {
+        throw MUJIN_EXCEPTION_FORMAT("failed to rewind inputStream (%s)", filename, MEC_InvalidArguments);
+    }
+    std::cout << contentLength << std::endl;
+
+    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_READFUNCTION, NULL, _ReadIStreamCallback);
     // prepare form
     struct curl_httppost *formpost = NULL;
     struct curl_httppost *lastptr = NULL;
     CURL_FORM_RELEASER(formpost);
     curl_formadd(&formpost, &lastptr,
                  CURLFORM_COPYNAME, "files[]",
```

Comment:
Maybe this should be a param of the function.

Reply:
Could you explain what do you mean by "this"?